### PR TITLE
Update job classAds from postJob

### PR DESCRIPTION
**Pull request by lecriste**

### CODE REVIEW

File: src/python/TaskWorker/Actions/PostJob.py [hunk 1]
```diff
@@ -2607,8 +2621,58 @@ def set_dashboard_state(self, state, reason = None, exitCode = None):
...
+        monitoringExitCode(params, exitCode)
+        msg += " ClassAds values to set are: %s" % (str(params))
+        self.logger.info(msg)
+        backendurls = self.server.get(self.rest_uri_no_api + '/info', data = {'subresource': 'backendurls'})[0]['result'][0]
+        loc = HTCondorLocator.HTCondorLocator(backendurls)
+        address = None
+        schedd = None
+        try:
+            self.logger.debug("Getting schedd object")
+            schedd, address = loc.getScheddObjNew(task['tm_schedd'])
```
Comment (lecriste) [inline src/python/TaskWorker/Actions/PostJob.py]: @belforte, probably this part is not necessary as the postJob runs on the schedd currently.
If at some point we move it on another machine, then we need this code. What's your opinion?

Comment (belforte) [reply src/python/TaskWorker/Actions/PostJob.py]: good point. Let me look at it carefull though. PostJos has a long history and I already tried and failed to clean it up one year ago. Don't break is our first priority.

Comment (belforte) [reply src/python/TaskWorker/Actions/PostJob.py]: @lecriste  seems you are referring to new code that you are introducing, not to existing code.
I am all for reviewing design of how to add the new functionalities, better if before writing code !

No. There is no way that we run the PostJob in any other place then the schedd. It is part of the DAG. YOu can take some functionality out of PostJob and run it elsewhere, if useful. But "some PostJob" will always be there as long as we have a DAG.
Maybe I miss something ?

File: src/python/TaskWorker/Actions/PostJob.py [hunk 2]
```diff
@@ -1513,6 +1517,7 @@ def execute(self, *args, **kw):
             self.logger.error(msg)
             if self.crab_retry is not None:
                 self.set_dashboard_state('FAILED')
+                self.set_postJob_ClassAds('FAILED')
```
Comment (belforte) [inline src/python/TaskWorker/Actions/PostJob.py]: I'd rather use a more descriptive name for this method, like self.report_state or self.send_state_to_monitor, or even self.set_state_classAd if we want to have an implementation dependent name

File: src/python/TaskWorker/Actions/PostJob.py [hunk 3]
```diff
@@ -2607,8 +2622,43 @@ def set_dashboard_state(self, state, reason = None, exitCode = None):
...
+        params = {'PostJobStatus': state}
+        self.monitoringExitCode(params, exitCode)
+        msg += " ClassAds values to set are: %s" % (str(params))
+        self.logger.info(msg)
+        schedd = htcondor.Schedd()
+        with HTCondorUtils.AuthenticatedSubprocess(os.environ['X509_USER_PROXY'], logger=self.logger) as (parent, rpipe):
+            if not parent:
+                for param in params:
+                    schedd.edit([self.dag_jobid], param, "'"+params[param]+"'")
+                # Allow Condor to remove the job from the queue before ten days upon completion
```
Comment (belforte) [inline src/python/TaskWorker/Actions/PostJob.py]: I fully do not understand this comment. English is somehow broken but mostly.. which ten days ? I do not see a ref. to ten days.

File: src/python/TaskWorker/Actions/PostJob.py [hunk 4]
```diff
@@ -2607,8 +2622,43 @@ def set_dashboard_state(self, state, reason = None, exitCode = None):
...
+        Update PostJobStatus and job exit-code among the job ClassAds for the monitoring script to update the Grafana dashboard.
+        """
+        if os.environ.get('TEST_POSTJOB_NO_STATUS_UPDATE', False):
+            return
+        msg = "postJob status: %s." % (state)
+        params = {'PostJobStatus': state}
+        self.monitoringExitCode(params, exitCode)
+        msg += " ClassAds values to set are: %s" % (str(params))
+        self.logger.info(msg)
+        schedd = htcondor.Schedd()
```
Comment (belforte) [inline src/python/TaskWorker/Actions/PostJob.py]: are you sure that this defaults to current schedd ? I am simply ignorant here.

Comment (lecriste) [reply src/python/TaskWorker/Actions/PostJob.py]: yes: https://htcondor-python.readthedocs.io/en/latest/htcondor.html#htcondor.Schedd.__init__

File: src/python/TaskWorker/Actions/PostJob.py (hunk 3 above)
Comment (lecriste) [reply src/python/TaskWorker/Actions/PostJob.py]: See the `LeaveJobInQueue` description here:
https://research.cs.wisc.edu/htcondor/manual/v7.6/10_Appendix_A.html

I agree the English is not clear :)

File: src/python/TaskWorker/Actions/PostJob.py (hunk 2 above)
Comment (lecriste) [reply src/python/TaskWorker/Actions/PostJob.py]: maybe `set_ClassAds_value`?

File: src/python/TaskWorker/Actions/DagmanCreator.py [hunk 5]
```diff
@@ -133,6 +133,10 @@
 WhenToTransferOutput = ON_EXIT_OR_EVICT
 +SpoolOnEvict = false
 
+# Keep job in the queue upon completion long enough for the postJob to run, allowing the monitoring script to fetch the postJob status and job exit-code updated by the postJob
+LeaveJobInQueue = true
```
[the remark this answers is not in the document]
Comment (lecriste) [reply src/python/TaskWorker/Actions/DagmanCreator.py]: there is no `+` sign, that's the Git syntax for new lines.

File: src/python/TaskWorker/Actions/DagmanCreator.py [hunk 6]
```diff
@@ -133,6 +133,10 @@
...
+PostJobStatus = "not run"
```
[the remark this answers is not in the document]
Comment (lecriste) [reply src/python/TaskWorker/Actions/DagmanCreator.py]: ok.

File: src/python/TaskWorker/Actions/PostJob.py (hunk 4 above)
Comment (belforte) [reply src/python/TaskWorker/Actions/PostJob.py]: I do not think that we use condor_submit -spool  In any case we want to define and enforce our own policy. 10 days are clearly too much.

File: src/python/TaskWorker/Actions/PostJob.py (hunk 3 above)
Comment (belforte) [reply src/python/TaskWorker/Actions/PostJob.py]: I do not think that we use condor_submit -spool  In any case we want to define and enforce our own policy. 10 days are clearly too much.

File: src/python/TaskWorker/Actions/DagmanCreator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (belforte) [reply src/python/TaskWorker/Actions/DagmanCreator.py]: oops  I misread the '+' from "add this line" as "a + char. in the code". sorry

File: src/python/TaskWorker/Actions/PostJob.py (hunk 2 above)
Comment (belforte) [reply src/python/TaskWorker/Actions/PostJob.py]: there are tons of ads areound, please have a name which makes it clear that only set one or two (at least IIUC).

**Comment by cronosnull** [comment]
It is possible to update the EnteredCurrentStatus with the time of the last change to the attribute? (thinking in the query to the history from the spider). 
If it is not possible, can we have another field with it?

**Comment by lecriste** [comment]
It is possible.
Let me check whether it may have any side effect.

**Comment by lecriste** [comment]
I decided to create a dedicated ClassAd, `CRAB_PostJobLastUpdate`, to not bother the Condor  `EnteredCurrentStatus`.
Jobs from task "190523_091243:lecriste_crab_NotAutomaticSplitting_data_preprod_LeaveJobInQueue_newPostJob_newDagmanCreator_LastUpdate" have it set to 0 by default and updated right before moving to the history.

**Comment by belforte** [comment]
+1

**Comment by cronosnull** [comment]
Is this change already deployed?

**Comment by lecriste** [comment]
Not in production.
I'm testing it in preprod so you can find several tasks of mine with this
change.

**Comment by cronosnull** [comment]
Ok, if you can send some tomorrow will be great (I deployed the change in the spider script some minutes ago).

**Comment by belforte** [comment]
I'll ran a few tasks now/tonight

**Comment by belforte** [comment]
Overall counts look OK so far, I will run more.
(I run two tasks which all failed due to using too much memory, would like to 
have a better mix).

I found one problem with status report. Look at this task
https://monit-grafana.cern.ch/d/cmsTMDetail/cms-task-monitoring-task-view?orgId=11&from=1559158715000&to=1559173853486&var-user=belforte&var-task=190529_213835%3Abelforte_crab_QCD_HT300to500_13TeV_8X_withPF-v1&var-site=All&var-schedd=0106&var-cmsUserId=cms1627&var-current_url=%2Fd%2FNvcuKTSiz%2Fcms_task_monitoring&var-task_str=belforte_crab_QCD_HT300to500_13TeV_8X_withPF-v1&var-status=Removed&var-crabid=All
And Job1.
From the PostJob log
https://cmsweb.cern.ch/scheddmon/0106/cms1627/190529_213835:belforte_crab_QCD_HT300to500_13TeV_8X_withPF-v1/postjob.1.0.txt
I'd expect it to be listed as failed with exit code 50660 . BUt instead is listed as Removed with exit code 80001 as per the job stdout
https://cmsweb.cern.ch/scheddmon/0106/cms1627/190529_213835:belforte_crab_QCD_HT300to500_13TeV_8X_withPF-v1/job_out.1.0.txt 
i.e. the reporting from the PostJob is missing.

About one half of jobs in this task show this prblem, while jobs in the other one, which all failed in the same way, are correctly labelled.

**Comment by belforte** [comment]
Comparing job 1 with e.g. job 5 in that task. There is a clear difference
I think it is the old thing that at times the jobs is killed by htcondor and we have all reporting, while if the out of memory is detected by the local site and the job killed there, the PostJob can't find the memory usage classAd and sets 50669 as exit code.
Yet in both cases we'd like the job to be listed as failed with an exit code.

Can you find if the problem is in the PostJob or in the spider or in MONIT ?

Another bad thing still for this task is that exit code count from old dashboard [1] is
50669 :  77 jobs
50660:  95 jobs
while in Grafana there are 75 and 97 

[1]
http://dashb-cms-job.cern.ch/dashboard/templates/task-analysis/#user=Stefano+Belforte+belforte%40infn.it&refresh=0&table=Jobs&p=1&records=200&sorting%5B%5D=3&sorting%5B%5D=desc&activemenu=2&status=&site=&tid=190529_213835%3Abelforte_crab_QCD_HT300to500_13TeV_8X_withPF-v1

the different ones are job #67 and #7

**Comment by belforte** [comment]
@lecriste I see that jobs in transferring are listed as Success in Grafana.
Is this a shortoming in our reporting from PostJob, or does grafana dashboard need to be modified to understand the new states ?

**Comment by lecriste** [comment]
> @lecriste I see that jobs in transferring are listed as Success in Grafana.
> Is this a shortoming in our reporting from PostJob, or does grafana dashboard need to be modified to understand the new states ?

The latter. Now it should work:
https://its.cern.ch/jira/browse/CMSMONIT-29?focusedCommentId=2634028&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2634028

**Comment by lecriste** [comment]
> I found one problem with status report. Look at this task
> https://monit-grafana.cern.ch/d/cmsTMDetail/cms-task-monitoring-task-view?orgId=11&from=1559158715000&to=1559173853486&var-user=belforte&var-task=190529_213835%3Abelforte_crab_QCD_HT300to500_13TeV_8X_withPF-v1&var-site=All&var-schedd=0106&var-cmsUserId=cms1627&var-current_url=%2Fd%2FNvcuKTSiz%2Fcms_task_monitoring&var-task_str=belforte_crab_QCD_HT300to500_13TeV_8X_withPF-v1&var-status=Removed&var-crabid=All
> And Job1.
> From the PostJob log
> https://cmsweb.cern.ch/scheddmon/0106/cms1627/190529_213835:belforte_crab_QCD_HT300to500_13TeV_8X_withPF-v1/postjob.1.0.txt
> I'd expect it to be listed as failed with exit code 50660 . BUt instead is listed as Removed with exit code 80001 as per the job stdout
> https://cmsweb.cern.ch/scheddmon/0106/cms1627/190529_213835:belforte_crab_QCD_HT300to500_13TeV_8X_withPF-v1/job_out.1.0.txt
> i.e. the reporting from the PostJob is missing.

I agree to list the job as `failed` instead of `removed`.
About the exit code, @cronosnull pointed out that, in presence of a `Chirp_CRAB3_Job_ExitCode`, it is preferred over `ExitCode`:
https://github.com/dmwm/cms-htcondor-es/blob/1268255eccd0525e5f31a98199fa4fa40675bdc0/src/htcondor_es/convert_to_json.py#L968-L970
Shall we change this prioritization? In all cases?
We looked for such a conversion in the old dashboard code and did not find any hint.

**Comment by lecriste** [comment]
@cronosnull, upon a discussion with Stefano we agreed that the first exit code to consider [here](https://github.com/dmwm/cms-htcondor-es/blob/1268255eccd0525e5f31a98199fa4fa40675bdc0/src/htcondor_es/convert_to_json.py#L968-L970) must be [the one set by the PotJob](https://github.com/lecriste/CRABServer/blob/c08e617ad5980fb34e5b11fd1591503094ad1b4e/src/python/TaskWorker/Actions/PostJob.py#L2670), namely `JobExitCode`.

**Comment by belforte** [comment]
I think we have a general guideline to prefix all classAds that we introduce with CMS_ , or CRAB_ if they are CRAB specific. Sorry that I did not point this out initially.

**Comment by belforte** [comment]
In any case, as this long saga testifies,there are so many places where an exit code is set/used/changed with various semantic etc. that a generic name like JobExtiCode (not to mention ExitCode) will be a source of headache when reviewing this 3 years from now.

**Comment by lecriste** [comment]
I did not add the prefix (yet) because that code was already there for the old dashboard. Changing the field names would probably break the old dashboard. 
Anyway, what about renaming `JobExitCode` and `ExeExitCode` [here](https://github.com/lecriste/CRABServer/blob/c08e617ad5980fb34e5b11fd1591503094ad1b4e/src/python/TaskWorker/Actions/PostJob.py#L2670-L2672) with `CRAB_PostJobExitCode` and `CMS_ExeExitCode` respectively?
I can create the PR (and a corresponding issue to not forget) to merge when the old dashboard will be decommissioned.

**Comment by belforte** [comment]
there should be no reason to user classAds which have same name as attributes sent to old dashboard.

**Comment by lecriste** [comment]
Indeed there is no reason other than avoid code duplication. I believe we will clean up all the old dashboard-related code once it is decommissioned.

**Pull request merged**